[mkdocs] docs: add transfer xem tutorial (#883)
NemProject/nem: mkdocs/pages/en/devbook/transactions/transfer-xem.md

### Fetching Network Time

{{ tutorial.code_snippet_tagged('step-3') }}

Every NEM transaction contains two time fields, both expressed in <network time:>,
the number of seconds since the NEM nemesis block:

* `timestamp`: The moment the transaction is created, set here to the current network time.
* `deadline`: How long the network keeps trying to confirm the transaction before discarding it.
    It must be after the timestamp and no more than 24 hours later.
    Otherwise, the node rejects the transaction.
    This example sets it two hours after the timestamp, well within the limit.

Building a transfer therefore needs an accurate network time.

The <get:/time-sync/network-time> endpoint reports the node's current network time.
The node returns this value in milliseconds, so the code divides it by 1000 to obtain the seconds that transactions
expect.

However, applications do not need to query the network time before every transaction.
It can be fetched once and then adjusted using the local system clock when needed.
This provides a good balance between accuracy and performance.

### Calculating the Transaction Fee

{{ tutorial.code_snippet_tagged('step-4') }}

> Every transaction pays a fee to the <harvester:> that includes it in a block.

Reviewer: ```suggestion
Every transaction pays a fee to the <harvester account:> that includes it in a block.
```

* `fee` is `fee_steps` multiplied by 50'000 atomic units, the value of one increment (0.05 XEM).

With the default 1 XEM, the fee falls in the first increment (0.05 XEM).

### Building the Transaction

{{ tutorial.code_snippet_tagged('step-5') }}

The snippet calls <dy:TransactionFactory.create> with a descriptor that supplies every required property of
the transfer transaction:

* `type`: This tutorial uses <ser:TransferTransactionV2>, the current transfer version, which can carry both XEM and
    other <mosaics:>.
    No mosaics are attached here, so the transaction sends XEM only.

* `signer_public_key`: The signer is the account that will pay the fee.
    In a transfer transaction, it is also the source of the transferred XEM.

* `fee`: The value calculated in the previous step. For 1 XEM, this is `50_000` atomic units (0.05 XEM).

* `timestamp` and `deadline`: The values computed in the network time step.

* `recipient_address`: The address that will receive the XEM.

* `amount`: The atomic-unit value computed earlier. For 1 XEM, this is `1_000_000`.

!!! info "Sending a mosaic or a message"

>     A <ser:TransferTransactionV2> can also carry other <mosaics:> instead of XEM, or include a <message:>, with the fee

Reviewer: The <message:> glossary entry is undefined. I guess you should add it in `transfer_transactions.md`.

* Create an <account:> to send the transfer transaction, either
    [from code](../accounts/create-from-private-key.md) or
    [by using a wallet](../../userbook/wallet/create-account.md).
* Obtain <XEM:> to pay for the transaction fee and transfer amount.
    See [Getting Testnet Funds from the Faucet](../accounts/testnet-faucet.md).

## Full Code

{% import 'tutorial.jinja2' as tutorial with context %}

{{ tutorial.code_full_tagged('devbook/transactions/transfer_xem', ['py', 'js']) }}

The whole code is wrapped in a single `try` block to provide simple error handling,
but applications will probably want to use more fine-grained control.

## Code Explanation

### Setting Up the Accounts

{{ tutorial.code_snippet_tagged('step-1') }}

Every transfer transaction involves two accounts: a **sender** and a **recipient**.

The **sender** is the <account:> that signs the transaction and pays the fee.
Its private key is loaded from the `SIGNER_PRIVATE_KEY` environment variable.
If not provided, a test key is used as default.

The **recipient** is the account that receives the XEM.
Its testnet <address:>, which is all that is needed to send funds, is loaded from the `RECIPIENT_ADDRESS`

> environment variable.

Author: ```suggestion
Its <address:> is loaded from the `RECIPIENT_ADDRESS` environment variable.
```

After a second read, there is no need to say testnet since it could work for mainnet too.

If not provided, a test address is used as default.

### Defining the Transfer Amount

{{ tutorial.code_snippet_tagged('step-2') }}

> The snippet defines the transfer amount in the `xem` variable, loaded from the `XEM_AMOUNT` environment variable.

Reviewer: The code assumes that `XEM_AMOUNT` is an integer, so the text should mention it.

Author: Good catch. I'm replacing it with a float since that seems like a common use case (e.g. transferring 1.5 XEM).

If not provided, a default of 1 XEM is used.

The transaction's `amount` field requires atomic units, not whole XEM.

> <XEM:> has a <divisibility:> of 6, so one XEM equals one million atomic units.

Reviewer: I think we've linked to XEM enough times already :D

* `deadline`: How long the network keeps trying to confirm the transaction before discarding it.
    It must be after the timestamp and no more than 24 hours later.
    Otherwise, the node rejects the transaction.
    This example sets it two hours after the timestamp, well within the limit.

Building a transfer therefore needs an accurate network time.

The <get:/time-sync/network-time> endpoint reports the node's current network time.
The node returns this value in milliseconds, so the code divides it by 1000 to obtain the seconds that transactions
expect.

However, applications do not need to query the network time before every transaction.
It can be fetched once and then adjusted using the local system clock when needed.
This provides a good balance between accuracy and performance.

### Calculating the Transaction Fee

{{ tutorial.code_snippet_tagged('step-4') }}

Every transaction pays a fee that rewards the nodes that process it and include it in a block.

NEM fees are not market-driven.
Instead of floating with network demand, NEM uses a fixed, published schedule, so the fee can be calculated in
advance without contacting a node.

> For an XEM transfer, the fee depends only on the amount being sent:

Reviewer: Ugh, I had no idea!
We definitely need to explain this in the Textbook, in a section about fees.
You can just move this description there and link it from here. And add the message cost, I guess.

And the regular transactions page in the Textbook also needs a section about fees :(

Author: Agree, and it gets way more complex when transferring non-XEM mosaics. I'm updating the textbook with all those learnings.

> For an XEM transfer, the fee depends only on the amount being sent:

Reviewer: It is now officially A XEM, not an XEM.
Don't ask me how to pronounce it, though.

* A transfer of up to 10,000 XEM costs 0.05 XEM.
* Above that, the fee rises by 0.05 XEM for every further 10,000 XEM sent.
* The fee is capped at 1.25 XEM, however large the transfer.

The snippet derives the fee from `xem`, defined in the previous step:

* `fee_steps` is the number of full 10'000-XEM increments in `xem`, clamped between 1 and 25.
* `fee` is `fee_steps` multiplied by 50'000 atomic units, the value of one increment (0.05 XEM).

With the default 1 XEM, the fee falls in the first increment (0.05 XEM).

Attaching a message or sending other <mosaics:> costs more than this XEM-only fee.

### Building the Transaction

{{ tutorial.code_snippet_tagged('step-5') }}

The snippet calls <dy:TransactionFactory.create> with a descriptor that supplies every required property of
the transfer transaction:

> * `type`: This tutorial uses `transfer_transaction_v2`, the current transfer version, which can

Reviewer: ```suggestion
* `type`: This tutorial uses <ser:TransferTransactionV2>, the current transfer version, which can
```

    carry both XEM and other <mosaics:>.
    No mosaics are attached here, so the transaction sends XEM only.

* `signer_public_key`: The signer is the account that will pay the fee.
    In a transfer transaction, it is also the source of the transferred XEM and mosaics.

* `fee`: The value calculated in the previous step. For 1 XEM, this is `50_000` atomic units (0.05 XEM).

* `timestamp` and `deadline`: The values computed in the network time step.

* `recipient_address`: The address that will receive the XEM.

* `amount`: The atomic-unit value computed earlier. For 1 XEM, this is `1_000_000`.

!!! info "Sending a mosaic or a message"

    A <ser:TransferTransactionV2> can also carry other <mosaics:> instead of XEM, or include a
    <message:>, with the fee calculated differently in each case.

>     See the [Transfer Mosaics](./transfer-mosaics.md) and [Transfer with a Message](./messages.md) tutorials.

Author: Links won't work until we create those tutorials.

The node validates the transaction as soon as it is announced and reports the outcome in the response.
A result of `SUCCESS` means the transaction passed this first check and was added to the <unconfirmed pool:>.
Any other result means the node did not accept it, and the response message explains why, for example that the
account does not hold enough XEM to cover the amount and the fee.

!!! warning "Do not rely on unconfirmed transactions"

    A `SUCCESS` result only means the transaction reached the unconfirmed pool.
    It is not yet guaranteed to be included in a block.
    Wait until it is [confirmed](#waiting-for-confirmation), and ideally past the <rewrite limit:>, before relying
    on it.

### Waiting for Confirmation

{{ tutorial.code_snippet_tagged('step-8') }}

The snippet above repeatedly queries the <get:/transaction/get> endpoint using the hash of the announced transaction.

!!! note "Polling vs WebSockets"

    This step uses polling to check whether the transaction has been confirmed.
    Polling is used here for illustration purposes, but it is not the recommended approach for real applications.

>     [WebSockets](../websockets/listen-transaction-flow.md) provide a more responsive solution without the overhead of

Author: Same, the link will break for now.
